client/app/counter/counter.component.ts

```diff
@@ -68,128 +55,64 @@ export class CounterComponent implements OnInit {
      */
     constructor(
         protected counterService: CounterService,
-        private formBuilder: FormBuilder,
+        private orderSocket: OrderSocketService
     ) { }
 
     ngOnInit() {
-        this.createForm();
-        this.calculateTotal();
+        this.getOrders();
+        this.configureSockets();
 
         this.dicountedValue = 0;
     }
 
     /**
-    * Creates the form
-    */
-    createForm(): void {
-        this.discountForm = this.formBuilder.group({
-            discount: this.formBuilder.control(0, [
-                Validators.required,
-                Validators.minLength(1),
-                Validators.maxLength(3),
-            ])
-        });
+     * Get all the orders from the database
+     */
+    getOrders() {
+        this.counterService
+            .getOrders()
+            .subscribe((data) => {
+                console.log(data);
+                this.orders = data;
+                this.sidebarOrder = data[0];
+            }, (error) => {
+            });
     }
-
     /**
-     * Calculate the total of all prices. Item price * quantity
+     * Listen for both the open and close methods to update table status on
+     * completion.
      */
-    calculateTotal() {
-        this.totalNoDiscount = this.order.reduce((accumulator, element) => element.price * element.quantity + accumulator, 0);
+    configureSockets() {
+        const openSub$ = this.orderSocket.getOrdersOpened();
+        const closedSub$ = this.orderSocket.getOrdersCompleted();
+
+        merge(openSub$, closedSub$)
+            .pipe(
+                tap(order => this.handleOrderEvent(order)),
+                untilDestroyed(this)
+            )
+            .subscribe();
     }
 
     /**
-     * Calculate  the discount out of the total price
+     * Handle locking a table when a order is added and unlocking when it is
+     * completed.
      */
-    addPromo() {
-        const discountValue = this.discountForm.value.discount;
-        this.dicountedValue = (discountValue / 100) * this.totalNoDiscount;
-
-        this.totalWithDiscount = (this.totalNoDiscount - this.dicountedValue);
+    handleOrderEvent(order) {
+        this.orders.push(order);
```

Comment:
This would currently add orders twice, once when created and then again when completed. Should check if its in the array already and update if so.

Reply:
The second param of the forEach callback is the index of the item `(element, index) =>` would remove the extra indexOf which is searching the array for the item again. Or reassigning this.orders with a `.filter(e => e._id !== order._id)` is another possible way to do it, which would remove the splice.